src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1195,17 +1463,735 @@ private static void ReportReadOnlyError(Symbol symbol, SyntaxNode node, BindValu
+
+                    return GetInvocationEscapeScope(
+                        indexerAccess.ReceiverOpt,
+                        indexerSymbol.Parameters,
+                        indexerAccess.Arguments,
+                        indexerAccess.ArgumentRefKindsOpt,
+                        indexerAccess.ArgsToParamsOpt,
+                        scopeOfTheContainingExpression);
+
+                case BoundKind.PropertyAccess:
+                    var propertyAccess = (BoundPropertyAccess)expr;
+
+                    // not passing any arguments/parameters
+                    return GetInvocationEscapeScope(
+                        propertyAccess.ReceiverOpt,
+                        default,
+                        default,
+                        default,
+                        default,
+                        scopeOfTheContainingExpression);
+            }
+
+            // At this point we should have covered all the possible cases for anything that is not a strict RValue.
+            return scopeOfTheContainingExpression;
+        }
+
+        /// <summary>
+        /// A counterpart to the GetRefEscape, which validates if given escape demand can be met by the expression.
+        /// The result indicates whether the escape is possible. 
+        /// Additionally, the method emits diagnostics (possibly more than one, recursively) that would help identify the cause for the failure.
+        /// </summary>
+        internal static bool CheckRefEscape(SyntaxNode node, BoundExpression expr, uint escapeFrom, uint escapeTo, bool checkingReceiver, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(!checkingReceiver || expr.Type.IsValueType || expr.Type.IsTypeParameter());
+
+            if (escapeTo >= escapeFrom)
+            {
+                // escaping to same or narrower scope is ok.
+                return true;
+            }
+
+            if (expr.HasAnyErrors)
+            {
+                // already an error
+                return true;
+            }
+
+            // void references cannot escape (error should be reported somewhere)
+            if (expr.Type?.GetSpecialTypeSafe() == SpecialType.System_Void)
+            {
+                return true;
+            }
+
+            // references to constants/literals cannot escape higher.
+            if (expr.ConstantValue != null)
+            {
+                Error(diagnostics, GetStandardRValueRefEscapeError(escapeTo), node);
+                return false;
+            }
+
+            switch (expr.Kind)
+            {
+                case BoundKind.ArrayAccess:
+                case BoundKind.PointerIndirectionOperator:
+                case BoundKind.PointerElementAccess:
+                    // array elements and pointer dereferencing are readwrite variables
+                    return true;
+
+                case BoundKind.RefValueOperator:
+                    // The undocumented __refvalue(tr, T) expression results in a variable of type T.
```

Comment:
I didn't find tests for __refvalue in the PR.

Reply:
We have a bunch of tests in ArglistTests, including some like RefValueUnsafeToReturn that try ref-returning them.

---
In reply to: [135401279](https://github.com/dotnet/roslyn/pull/21334#discussion_r135401279) [](ancestors = 135401279)

Reply:
Will add a test for val-escaping of a span that wraps a __refvalue. That would be the only observable scenario that is new. It is quite a corner case though.

---
In reply to: [135663930](https://github.com/dotnet/roslyn/pull/21334#discussion_r135663930) [](ancestors = 135663930,135401279)